ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -339,37 +339,36 @@ private void setEndOffsets() throws IOException {
+   * Atomically adds {@code segmentFilesToAdd} to and removes {@code segmentsToRemove} from the map of {@link Offset} to
    * {@link IndexSegment} instances.
-   * @param filesToAdd the backing files of the {@link IndexSegment} instances to add.
+   * @param segmentFilesToAdd the backing files of the {@link IndexSegment} instances to add.
    * @param segmentsToRemove the start {@link Offset} of {@link IndexSegment} instances to remove.
-   * @throws IllegalArgumentException if any {@link IndexSegment} that needs to be added has an offset higher than that
-   * in the journal.
-   * @throws StoreException if an {@link IndexSegment} instance cannot be created for a provided segment to add.
+   * @throws IllegalArgumentException if any {@link IndexSegment} that needs to be added or removed has an offset higher
+   * than that in the journal.
+   * @throws StoreException if an {@link IndexSegment} instance cannot be created for the provided files to add.
    */
-  void changeIndexSegments(List<File> filesToAdd, Set<Offset> segmentsToRemove) throws StoreException {
+  void changeIndexSegments(List<File> segmentFilesToAdd, Set<Offset> segmentsToRemove) throws StoreException {
     TreeMap<Offset, IndexSegment> segmentsToAdd = new TreeMap<>();
-    for (File indexSegmentFile : filesToAdd) {
+    Offset journalFirstOffset = journal.getFirstOffset();
+    for (File indexSegmentFile : segmentFilesToAdd) {
       IndexSegment indexSegment = new IndexSegment(indexSegmentFile, true, factory, config, metrics, journal);
+      if (indexSegment.getEndOffset().compareTo(journalFirstOffset) > 0) {
+        throw new IllegalArgumentException("One of the index segments has an end offset " + indexSegment.getEndOffset()
+            + " that is higher than the first offset in the journal " + journalFirstOffset);
+      }
       segmentsToAdd.put(indexSegment.getStartOffset(), indexSegment);
     }
 
     indexesChangeLock.lock();
     try {
-      if (segmentsToAdd.size() > 0) {
-        Offset lastSegmentToAddEndOffset = segmentsToAdd.lastEntry().getValue().getEndOffset();
-        Offset journalFirstOffset = journal.getFirstOffset();
-        if (lastSegmentToAddEndOffset.compareTo(journal.getFirstOffset()) > 0) {
-          throw new IllegalArgumentException("One of the index segments has an end offset " + lastSegmentToAddEndOffset
-              + " that is higher than the first offset in the journal " + journalFirstOffset);
-        }
-      }
-
       ConcurrentSkipListMap<Offset, IndexSegment> oldIndexSegments = indexesRef.get();
       ConcurrentSkipListMap<Offset, IndexSegment> newIndexSegments = new ConcurrentSkipListMap<>();
       for (Map.Entry<Offset, IndexSegment> entry : oldIndexSegments.entrySet()) {
         if (!segmentsToRemove.contains(entry.getKey())) {
           newIndexSegments.put(entry.getKey(), entry.getValue());
+        } else if (entry.getKey().compareTo(journal.getFirstOffset()) >= 0) {
```

Comment:
1. The last index segment can have the start offset that lies within the journal right? Why is this check required. 2. This was bit confusing. Can you have two diff loops. One loop to go over segmentsToRemove and throw if any of the start offsets lie within journal. 2nd loop to create the newIndexSegments. bcoz, if you want to, you can combine the previous check (segmentFilesToAdd's ) in here itself. But I feel having it separately makes it more readable.